### ISSUE TITLE
Can't figure out how to use EFCoreStore and MultiTenantIdentity at the same time

### ISSUE THREAD

**Issue by dwt12777** [Finbuckle/Finbuckle.MultiTenant#306]
I'm trying to set up my project so that:

1. App data and tenant data is stored in a SQL server db via Entity Framework
2. The app should use the multitenant version of Identity so that each tenant can set up and maintain their own users

And and can't figure out how to get it done.  I finally got the first point working (I think).  I got everything set up finally, and added a migration to get ready to update the database structures.  I was super excited to look at the migration and see that it automatically picked up the fact that it needs to build a tenantinfo table... but then I was really confused to see why the migration wants to add that table but also drop all the Identity tables!  (They were set up as part of the original project template before I started playing with Finbckle.MultiTenant).

In order to get to this point, I had to make sure my AppDbContext was set up like the efcoredatastore example `public class AppDbContext : EFCoreStoreDbContext<TenantInfo>` - but that is different than the Identity example where the AppDbContext inherits (or implements?  I get those confused) the class: `MultiTenantIdentityDbContext`.  I'm wondering if the database migrations want to strip all the identity tables because I'm not using the MultiTenantIdenityDbContext, but I can't use it because I also need to use the EFCoreStoreDbContext!

In other words, I feel like I need to use EFCoreStoreDbContext and MultiTenantIdenityDbContext at the same time but it seems like I need to pick one or the other.  Is there a way to do both?

Hope this makes sense - any suggestions?

**Comment by svenkle** [comment]
@dwt12777 the key problem here is that you can't inherit from both data stores (no multiple inheritance in C# sadly) and they both extend from different bases. I've recently been working on solving a similar issue and believe that conceptually the tenant data store and the application data store are different - it just so happens you want to store two distinct sets of data in the same database. 

Three main ideas came to mind when investigating:

- Submitting a PR where the `WithEFCoreStore()` extension and it's underlying stores are updated to support a custom context providing it inherits from a given interface and/or base.

- Use the existing `EFCoreStore` and just configure it to point at the same underlying database.

- Create my own `IMultiTenantStore<>` which just happens to store it's data in my `ApplicationDbContext` under the hood.

For the sake of speed (timelines and deadlines), I opted for the third option. At the time of writing, I haven't tested the solution below to it's full extent however, database migrations are working and overall the application appears to function as expected.

1. Create a `IMultiTenantStore` class `ApplicationDbContextStore` that matches almost identically to the standard `EFCoreStore` class. The key differences being slightly different generic parameters and an updated constructor to take the custom `ApplicationDbContext`.

2. Add a `DbSet<>` for the TenantInfo records onto `ApplicationDbContext` eg. `DbSet<TenantInfo> TenantInfos`. This can be called anything however, "TenantInfos" is as close as it gets to the `EFCoreStore` implementation without the name collision `TenantInfo` is already in use through the `IMultiTenantDbContext` interface.

3. Override or extend the `OnModelCreating` for `ApplicationDbContext` to ensure that the same fluent configuration from the `EFCoreStore` class is applied to the `TenantInfos` DbSet. In v5 the constants for max length are public, however in v6 you must replace the constant with your own. eg. 64.

4. Use the standard `WithStore<>()` method eg. `
				.WithStore<ApplicationDbContextStore<TenantInfo>>(ServiceLifetime.Scoped)` to configure your new store.

With a few small adjustments to this code base it should be easy enough to achieve custom contexts without having to write a custom store (although its very easy to do so). This would probably warrant a discussion with @AndrewTriesToCode as I'm not 100% happy with the solution above.

Hope this helps.

**Comment by dwt12777** [comment]
Thank you @svenkle - I ended up writing a custom store (saves tenants to a JSON file, probably kind of like the config store, but as I understood it - that was read only and I wanted a store that I could read/write to...)

But anyway, you were right, once I did that, I was was able to use the `IMultiTenantDbContext`... I can't pretend like I understand it all yet, but I can tell you the app now has identity tables with tenantIds in my sql database... progress!

**Comment by AndrewTriesToCode** [comment]
hi @dwt12777 I'm glad you found a good alternative.

@svenkle thanks for assisting with the detailed response

In general I recommend having separate multitenant store db contexts and main application db contexts (although this isn't true in all cases). I realize my sample programs aren't very good about this and need to improve them.

**Comment by morindo** [comment]
> Thank you @svenkle - I ended up writing a custom store (saves tenants to a JSON file, probably kind of like the config store, but as I understood it - that was read only and I wanted a store that I could read/write to...)
> 
> But anyway, you were right, once I did that, I was was able to use the `IMultiTenantDbContext`... I can't pretend like I understand it all yet, but I can tell you the app now has identity tables with tenantIds in my sql database... progress!

Greetings @dwt12777, do you think you could share your JSON file store read/write configuration store code?
